File: day03/main.cpp
#include <iostream>
#include <fstream>
#include <vector>
#include <algorithm>
#include <numeric>
#include <set>

auto read()
{
    std::fstream ifs(std::string(SRC_DIR)+"/input.txt");
    std::string line;
    std::vector<std::string> result;
    while(std::getline(ifs, line))
    {
        if(line.empty())
            continue;
        result.push_back(line);
    }
    return result;
}

int64_t convert(char c)
{
    if(c>='a' && c<='z')
    {
        return (c-'a'+1);
    }

    assert(c>='A' && c<='Z');
    return (c-'A'+27);
}

void part1()
{
    int64_t result = 0;
    for(const auto& line: read())
    {
        const int len = line.size();
        std::set<char> first(line.begin(), line.begin()+len/2);
        std::set<char> second(line.begin()+len/2, line.end());

        std::vector<char> intersection;
        std::set_intersection(first.begin(), first.end(),second.begin(), second.end(), std::back_inserter(intersection));
        assert(intersection.size()==1);
        const char c = intersection.front();
        result+=convert(c);
    }
    std::cout << "Part1: " << result << std::endl;
}

void part2()
{
    int64_t result = 0;
    const auto lines = read();

    for(int i=0; i < lines.size(); i+=3)
    {
        std::string first = lines[i];
        std::string second = lines[i+1];
        std::string third = lines[i+2];
        std::sort(first.begin(), first.end());
        std::sort(second.begin(), second.end());
        std::sort(third.begin(), third.end());

        std::string first_intersect;
        std::set_intersection(first.begin(), first.end(),second.begin(), second.end(), std::back_inserter(first_intersect));

        std::string second_intersect;
        std::set_intersection(third.begin(), third.end(),first_intersect.begin(), first_intersect.end(), std::back_inserter(second_intersect));

        result+=convert(*second_intersect.begin());
    }
    std::cout << "Part2: " << result << std::endl;
}

int main() {

    part1();
    part2();
    return 0;
}

File: day03/src/main.rs
use std::collections::HashSet;
use std::fs::File;
use std::io::{self, BufRead};
use std::path::Path;


fn read_lines<P>(filename: P) -> io::Result<io::Lines<io::BufReader<File>>> where P: AsRef<Path> {
    let file = File::open(filename)?;
    Ok(io::BufReader::new(file).lines())
}

fn read() -> Vec<String> {
    let file_path = "input.txt";
    let Ok(lines) = read_lines(file_path) else {
        return vec!();
    };

    return lines.into_iter().map(|x| x.unwrap()).collect();
}

fn score(c: char) -> i32 {
    if c >= 'a' && c <= 'z' {
        return (c as i32) - ('a' as i32) + 1;
    }
    assert!(c >= 'A' && c <= 'Z');
    return (c as i32) - ('A' as i32) + 27;
}

fn part1() {
    let result: i32 = read().into_iter().map(|s| {
        let mut chunks = s.as_bytes().chunks(s.len() / 2);
        let first = chunks.next().unwrap().iter().cloned().collect::<HashSet<u8>>();
        let second = chunks.next().unwrap().iter().cloned().collect::<HashSet<u8>>();
        let intersection = *first.intersection(&second).next().unwrap();
        return score(intersection as char);
    }).sum();

    println!("Part1: {}", result)
}

fn part2() {
    let result: i32 = read().chunks(3).into_iter().map(|s| {
        assert_eq!(s.len(), 3);
        let first = s[0].as_bytes().iter().cloned().collect::<HashSet<u8>>();
        let second = s[1].as_bytes().iter().cloned().collect::<HashSet<u8>>();
        let third = s[2].as_bytes().iter().cloned().collect::<HashSet<u8>>();

        let mut intersection = first.intersection(&second).cloned().collect::<HashSet<u8>>();
        intersection = third.intersection(&intersection).cloned().collect::<HashSet<u8>>();
        assert_eq!(intersection.len(), 1);
        return score(*intersection.iter().next().unwrap() as char);
    }).sum();

    println!("Part2: {}", result)
}

fn main() {
    part1();
    part2();
}


